ten-simple-rules-dockerfiles.Rmd

```diff
@@ -562,70 +561,64 @@ RUN pip install \
-If you developed extensive software for a specific analysis, you should ideally publish it as a software package and follow \ruleref{rule:pinning} for installing it (cf. "A package first" in @gruening_recommendations_2019).
-Consider using a suitable package system with pinned versions even for small scripts if the functions are reusable across datasets or workflows by you or others.
-If you cannot publish a software project in the appropriate language's package structure, you should make sure the source code files you `COPY` into the image are published together with the `Dockerfile` (see&nbsp;\ruleref{rule:publish}).
-If the code and `Dockerfile` are in the same repository, the origin and results of the `COPY` instruction and subsequent installation code will be understandable in the future.
-
-Storing _data files_ outside of the container allows for handling of very large datasets and datasets with data worthy of protection, e.g., proprietary data or private information.
-Do not include such data in an image.
+Storing _data files_ outside of the container allows handling of very large or sensitive datasets, e.g., proprietary data or private information.
+Do not include such data in an image!
 To avoid publishing sensitive data by accident, you can add the data directory to the [`.dockerignore`](https://docs.docker.com/engine/reference/commandline/build/#use-a-dockerignore-file) file, which excludes files and directories from the [build context](https://docs.docker.com/engine/reference/commandline/build/#extended-description), i.e., the set of files considered by `docker build`.
-Ignoring data files also speeds up the build in cases where there are very large files or many small files; for example, large data files or the `.git` folder of the [Git version control system](https://git-scm.com/), which contains many small files across numerous folders, are usually not needed during image build.
-In any case, you should include dummy or test data into the image to ensure that a container is functional without the actual dataset, e.g., for automated tests, instructions in the user manual, or peer review (see also "functional testing logic" in @gruening_recommendations_2019).
-For all these cases, you should provide clear instructions for users in the `README` on how to obtain actual or dummy data.
-When publishing your workspace, e.g., on Zenodo, having data and script contents as regular files outside of the container makes them more accessible to others, for example for reuse or analysis.
+Ignoring data files also speeds up the build in cases where there are very large files or many small files.
+You should include dummy or test data in the image to ensure that a container is functional without the actual dataset, e.g., for automated tests, instructions in the user manual, or peer review (see also "functional testing logic" in @gruening_recommendations_2019).
```

Comment:
Ah here we go! I wonder how this can be referred to or alluded to earlier, so we don't basically say "Don't put any data in your container!" but then say "well, small or dummy is ok ;)"

Reply:
Well we could lead with the (general) rule and then mention an exception later (most rules have them after all!) or we could revise the rule to be "Mount *real* datasets at run time" and rewrite accordingly. I think the former makes for simpler, stronger messaging when you imagine that some fraction of readers will probably only look at rule headings (and perhaps the first sentence or two of each). I've edited to flag the later exception but can change it if people prefer the latter option.

Reply:
I think the general rule and the exception are clear and understandable.